### PR TITLE
Redirect graceful exiting to standard exit flow

### LINKED ISSUE #10630: ConfirmExitDialog interferes with auto exit when installing an update
**Describe the bug:** 
On Windows, after downloading a new release, clicking the `UpdateCompleteNotification` 
("Update ready to install. Click to restart!") will start a 2 second exit timer,
but it also triggers a `ConfirmExitDialog`.

This creates a weird UX issue where the user can either
* click "Goodbye" -> exit right away
* click "Just a little more" -> asks again 2 seconds later
* take more than 2 seconds to act -> exits automatically

I don't see the point in showing a dialog if the user cannot abort either way,
although the method doc seems to suggest this is somewhat intended... (?)
https://github.com/ppy/osu/blob/ea007af57260f0a0121c57c2d2d4aaad7bb46730/osu.Game/OsuGame.cs#L925-L929

**osu!lazer version:** 
2020.1031.0 (Windows)

### PR DESCRIPTION
**Pull request by peppy**
Rather than doing a completely forced exit as we were, this will now follow a more standard flow with the ability for the user to abort along the way. This is more in line with how I wanted this to work.

Note that this means a confirmation is now shown. It has been discussed in the past that this confirmation should only show when there's an ongoing action implies the user may want to cancel the exit. For now I think this is fine.

Addresses https://github.com/ppy/osu/discussions/18399#discussioncomment-2811311.
Closes #10630.

### CODE REVIEW

File: osu.Game/OsuGameBase.cs [hunk 1]
```diff
@@ -419,7 +419,7 @@ public override void SetHost(GameHost host)
         /// Use to programatically exit the game as if the user was triggering via alt-f4.
         /// Will keep persisting until an exit occurs (exit may be blocked multiple times).
```
Comment (frenzibyte) [inline osu.Game/OsuGameBase.cs:420]: Would reword to give the idea that `GracefullyExit` could be overriden, or maybe remove this line entirely.
```suggestion
        /// By default, this will keep persisting until an exit occurs (exit may be blocked multiple times).
```

**Review by frenzibyte — DISMISSED** [review]

**Comment by bdach** [comment]
Does this close https://github.com/ppy/osu/issues/10630?

**Comment by peppy** [comment]
It should in a way, yeah. It will mean the user *can* now abort the update process, but I think that's fine for now, as per OP.

**Comment by bdach** [comment]
`GracefullyExit()` has more consumers, though, like `MigrationSelectScreen` - doesn't this change potentially break them?

**Comment by peppy** [comment]
It won't break anything, but will give the user a chance to abort. For the migration one I agree it's a touch weird, but likely better than what we had. My thinking is that if there's a reason to require confirmation of any kind, all quit prompts should hit it.

For migration, the worst that can happen is the user creates files at the old storage location which are discarded afterwards.

**Comment by bdach** [comment]
Seems really weird to have a popup open that says this:

https://github.com/ppy/osu/blob/3105e04839c87acf175fd4623d647fdd15da905d/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs#L56

and then ask for confirmation. But alright, let's see if anyone manages to shoot their foot off with this.

**Comment by peppy** [comment]
We can change that migration flow to use a direct host.Exit call if you find that better. Will need to test if it works though.

**Comment by peppy** [comment]
Also please keep in mind what I said in the OP: the confirmation should only appear if there’s a reason for it to (ongoing actions the it would interrupt). In the majority of cases Inwould expect it to not show, once that is improved.

**Comment by bdach** [comment]
> the confirmation should only appear if there’s a reason for it to (ongoing actions the it would interrupt)

Well I'm testing the migration flow right now on windows and the exit confirmation dialog shows for me every time without fail. And I haven't initiated any imports or anything of the sort, just triggering the migration immediately after launch.

**Comment by bdach** [comment]
Ok I think I misunderstood intent - the "should" in the messages above is an "eventually should" rather than a "it should already", I guess?

Would prefer to fix the migration screen usage so that it actually exits without being able to block it and have it do weird stuff but not hugely bothered. Up to you.

**Comment by peppy** [comment]
> Ok I think I misunderstood intent - the "should" in the messages above is an "eventually should" rather than a "it should already", I guess?

Yep, requires some future work.
 
> Would prefer to fix the migration screen usage so that it actually exits without being able to block it and have it do weird stuff but not hugely bothered. Up to you.

Have changed this flow and tested that it works as expected still.

**Review by bdach — DISMISSED** [review]

**Review by bdach — APPROVED** [review]
